streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -982,17 +1000,34 @@ private synchronized void setStateWhenNotInPendingShutdown(final State newState)
         setState(newState);
     }
 
-    private synchronized void setState(final State newState) {
-        final State oldState = state;
-        if (!state.isValidTransition(newState)) {
-            log.warn("{} Unexpected state transition from {} to {}.", logPrefix, oldState, newState);
-        } else {
-            log.info("{} State transition from {} to {}.", logPrefix, oldState, newState);
-        }
+    /**
+     * Sets the state
+     * @param newState New state
+     */
+    void setState(final State newState) {
+        synchronized (stateLock) {
+            final State oldState = state;
+
+            // there are cases when we shouldn't check if a transition is valid, e.g.,
+            // when, for testing, a thread is closed multiple times. We could either
+            // check here and immediately return for those cases, or add them to the transition
+            // diagram (but then the diagram would be confusing and have transitions like
+            // PENDING_SHUTDOWN->PENDING_SHUTDOWN).
+            if (newState != State.DEAD && (state == State.PENDING_SHUTDOWN || state == State.DEAD)) {
```

Comment:
@enothereska could you elaborate a bit more here? More concretely, could you update the comment on top to list all the transitions that you were trying to bypass with this condition so we can discuss if they make sense case-by-case:

```
            // there are cases when we shouldn't check if a transition is valid, e.g.,
            // when, for testing, a thread is closed multiple times. We could either
            // check here and immediately return for those cases, or add them to the transition
            // diagram (but then the diagram would be confusing and have transitions like
            // PENDING_SHUTDOWN->PENDING_SHUTDOWN).
```